### PR TITLE
Feature oauth2

### PR DESCRIPTION
**Pull request by m4github**
OAuth component added to the project and clean the Spotify component.

### CODE REVIEW

**Comment by MohammadAbbasi76** [comment]
please use the same rule for naming, like other source files.

File: components/OAuth2.0/authorization.c [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        return false;
+    }
+
+    cJSON *accessTokenObj = cJSON_GetObjectItem(J_Token, "access_token");
+    if (accessTokenObj != NULL && accessTokenObj->type == cJSON_String) 
+    {
+        strncpy(token->AccessToken, accessTokenObj->valuestring, ACCESS_TOKEN_STR_SIZE - 1);
+        token->AccessToken[ACCESS_TOKEN_STR_SIZE - 1] = '\0';
+    }
+
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: All Json function should be in special source file like Spotify source!!

File: components/OAuth2.0/authorization.c [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+/**
+ * @brief This function reads refresh token from spiffs and send request for new token
+ * @return True if token received and saved, false for otherwise
+ */
+static bool Oauth_TokenRenew(void)
+{
+    char receivedData[LONG_BUF];
+    ReadTxtFileFromSpiffs(pInterfaceHandler->ConfigAddressInSpiffs,
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: "pInterfaceHandler" this is global variable, please use the Pascal way for naming.

**Comment by MohammadAbbasi76** [comment]
please use the same naming in "OauthHttpLocalServer.c" file, at now all functions doesn't follow same rule.
Screenshot from 2024-07-07 11-07-22

**Comment by MohammadAbbasi76** [comment]
why you remove "StopWebServer" function in OauthHttpLocalServer.c!!! 
please write it again.

**Comment by MohammadAbbasi76** [comment]
If you don't need to some source or header file please remove them!
(in Spotify component)
Screenshot from 2024-07-07 11-12-37

**Review by MohammadAbbasi76 — CHANGES_REQUESTED** [review]
please make a struct and give all parameters from the user, you use HARD CODE all where, what was your main duty?

File: components/OAuth2.0/OauthAPICall.c [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ * @param[in] SizeCode The size of the authorization code.
+ * @return This function does not return a value.
+ */
+void SendTokenRequest(char *Code)
+{  
+    esp_http_client_config_t clientConfig = {
+        .url = "https://accounts.spotify.com/api/token",                            
+        .host = "accounts.spotify.com",
+        .path = "/api/token",
+        .method = HTTP_METHOD_POST,
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/OauthAPICall.c]: all where you use HARD CODE!

File: components/OAuth2.0/OauthAPICall.c [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * @brief This function sends a request to the Service login API to exchange an authorization code for an access token.
+ * @return This function does not return a value.
+ */
+void SendRequest_ExchangeTokenWithRefreshToken(char *RefreshToken)
+{
+    esp_http_client_config_t clientConfig = {
+        .url = "https://accounts.spotify.com/api/token",
+        .host = "accounts.spotify.com",
+        .path = "/api/token",
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/OauthAPICall.c]: take from user!

**Comment by MohammadAbbasi76** [comment]
Your program doesn't work, please fix it!
Screenshot from 2024-07-07 11-24-44

File: components/OAuth2.0/include/OauthHttpLocalServer.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+#ifndef OAUTH_HTTP_SERVER
+#define OAUTH_HTTP_SERVER
+
+#include "authorization.h"
+
+extern QueueHandle_t SendCodeFromHttpToTask;
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/include/OauthHttpLocalServer.h:10]: please use pointer instead "extern"

File: components/OAuth2.0/authorization.c [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+#include <stdlib.h>
+#include "authorization.h"
+#include "OauthHttpLocalServer.h"
+#include "SpiffsManger.h"
+#include "OauthAPICall.h"
+#include "cJSON.h"
+
+// ****************************** Global Variables //FIXME merge structs
+ServiceInterfaceHandler_t *pInterfaceHandler;
+ServiceInterfaceHandler_t ServiceInterfaceHandler;
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: we have a service manager in another project with this name. please use the unique name related to your component!

File: components/OAuth2.0/authorization.c [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+                }
+                SendTokenRequest(receivedData);                                                         // send request for Token
+                OAuthInterfaceHandler.Status = AUTHENTICATED;                                                          // Code received and checked, so update Status to AUTHENTICATED         
+                break;
+            }
+            case AUTHENTICATED:
+            {
+                ESP_LOGI(TAG, "AUTHENTICATED");
+                bool IsServiceResponded = xSemaphoreTake(OAuthInterfaceHandler.OAuthBuffer.ResponseReadyFlag, pdMS_TO_TICKS(SEC));              // Waiting for Token to be recieved by queue
+                if (!IsServiceResponded)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: you don't create semaphore, this is the reason that your program doesn't work.

File: components/OAuth2.0/authorization.c [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    ESP_LOGI(TAG, "new Token found!");
+    OAuthInterfaceHandler.TokenLastUpdate = xTaskGetTickCount();
+    return true;
+}
+
+static void Oauth_MainTask(void *pvparameters)
+{
+    while (1)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: please use "true" instead of 1 in the while

File: components/OAuth2.0/authorization.c [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+#include <stdlib.h>
+#include "authorization.h"
+#include "OauthHttpLocalServer.h"
+#include "SpiffsManger.h"
+#include "OauthAPICall.h"
+#include "cJSON.h"
+
+// ****************************** Global Variables //FIXME merge structs
+OAuthInterfaceHandler_t *pInterfaceHandler;
+OAuthInterfaceHandler_t OAuthInterfaceHandler;
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: you need private hander, please use this method "SpotifyInterfaceHandler_t *InterfaceHandler;
SpotifyPrivateHandler_t PrivateHandler;"

**Comment by m4github** [comment]
> why you remove "StopWebServer" function in OauthHttpLocalServer.c!!! please write it again.

because it was not mentioned anywhere so I commented it as temporary

File: components/OAuth2.0/authorization.c (hunk 7 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: It was created in taskinit func.

File: components/OAuth2.0/authorization.c (hunk 9 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: We talked about it and Private-handler is not needed.

File: components/OAuth2.0/authorization.c (hunk 1 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: It's a static function called just for ones so I prefer to keep it here in *.c file...It looks more optimal this way.

File: main/main.c [hunk 10]
```diff
@@ -2,72 +2,89 @@
...
+#include "esp_http_client.h"
+
 #define TIMER_TIME pdMS_TO_TICKS(500) // in millis
 
 // ****************************** GLobal Variables ****************************** //
-static const char *TAG = "Main";
+//static const char *TAG = "Main";
+OAuthInterfaceHandler_t OAuthInterfaceHandler;
 SpotifyInterfaceHandler_t SpotifyInterfaceHandler;
+HttpInfo_t ClientInfo;
```
Comment (Behn4m) [inline main/main.c]: name is not good. it should start with Spotify_

File: components/OAuth2.0/authorization.c [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+/**
+ * @brief This function initiates the authorization process.
+ * @param InterfaceHandler as the handler
+ * @return true if task run to the end
+ */
+bool Oauth_TaskInit(OAuthInterfaceHandler_t *InterfaceHandler)
+{
+    AuthInterfaceHandler = InterfaceHandler;
+    AuthInterfaceHandler->Status = INIT;
```
Comment (Behn4m) [inline components/OAuth2.0/authorization.c]: the handler that holds private variables such as `Status` should be same as the handler that is exposed in the API header file.

File: components/OAuth2.0/authorization.c [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+            "Oauth_MainTask",                  // Task name (for debugging)
+            SERVICE_TASK_STACK_SIZE,             // Stack size (in words)
+            NULL,                                // Task parameters (passed to the task function)
+            tskIDLE_PRIORITY + SERVICE_PRIORITY, // Task priority (adjust as needed)
+            xStack,                              // Stack buffer
+            xTaskBuffer                          // Task control block
+        );
+
+        // Allocate buffer and Initialize the service API call //TODO change dynamic mem to static
+        AuthInterfaceHandler->OAuthBuffer.MessageBuffer =
```
Comment (Behn4m) [inline components/OAuth2.0/authorization.c]: same comment. this buffer is private, should not be exposed

File: components/OAuth2.0/authorization.c [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+                (char *)malloc(SUPER_BUF * sizeof(char));    
+        AuthInterfaceHandler->OAuthBuffer.ResponseReadyFlag = xSemaphoreCreateBinary();
```
Comment (Behn4m) [inline components/OAuth2.0/authorization.c]: private variable is exposed to the main!

File: components/OAuth2.0/OauthHttpLocalServer.c [hunk 14]
```diff
@@ -146,15 +145,15 @@ httpd_handle_t Spotify_StartWebServer()
...
+// {
+//     return httpd_stop(HttpdServerHandler);
+// }
 
 /**
  * @brief This function starts the mDNS service.
  */
-bool Spotify_StartMDNSService()
+static bool Oauth_StartMDNSService()
 {
```
Comment (Behn4m) [inline components/OAuth2.0/OauthHttpLocalServer.c]: mdns `hostname` should be an input of this function

File: components/OAuth2.0/OauthHttpLocalServer.c [hunk 15]
```diff
@@ -188,3 +187,29 @@ bool Spotify_StartMDNSService()
     ESP_LOGI(TAG, "MDNS Init done");
     return true;
 }
+
+/**
+ * @brief Run Http local service
+ */
+bool HttpServerServiceInit()
+{
```
Comment (Behn4m) [inline components/OAuth2.0/OauthHttpLocalServer.c]: you should find a way to know the mdns `hostname` here to pass it the `Oauth_StartMDNSService` function

File: components/OAuth2.0/OauthHttpLocalServer.c [hunk 16]
```diff
@@ -188,3 +187,29 @@ bool Spotify_StartMDNSService()
...
+    }
+    
+    bool IsMdnsStarted = Oauth_StartMDNSService();
+    if (!IsMdnsStarted)
+    {
+        ESP_LOGE(TAG, "Running mDNS failed!");
+        return false;
+    };
+
+    ESP_LOGI(TAG, "** local server created, mDNS is running! **");
```
Comment (Behn4m) [inline components/OAuth2.0/OauthHttpLocalServer.c]: it is a good idea to print the mdns `hostname` in the log!

File: components/OAuth2.0/OauthHttpLocalServer.c [hunk 17]
```diff
@@ -102,26 +101,26 @@ static esp_err_t Spotify_HttpsCallbackHandler(httpd_req_t *HttpdRequest)
...
     .method = HTTP_GET,
-    .handler = Spotify_RequestDataAccess};
+    .handler = RequestDataAccess};
 
 /**
  * this strcut is http URL handler if receive "/callback" HttpsUserCallBackFunc getting run
  */
-static const httpd_uri_t Spotify_Response_Access_URI = {
+static const httpd_uri_t Response_Access_URI = {
     .uri = "/callback/",
```
Comment (Behn4m) [inline components/OAuth2.0/OauthHttpLocalServer.c]: the `uri` should also be need to be configurable

File: components/OAuth2.0/OauthHttpLocalServer.c [hunk 18]
```diff
@@ -44,11 +43,11 @@ bool Spotify_FindCode(char *Response, uint16_t SizeRes)
 }
 
 /**
- * @brief This function handles the first HTTPS request to Spotify and redirects the user to the authorization page.
+ * @brief This function handles the first HTTPS request to Service and redirects the user to the authorization page.
  * @param[in] req The HTTP request object.
  * @return Returns ESP_OK if the request is processed successfully.
  */
-static esp_err_t Spotify_RequestDataAccess(httpd_req_t *HttpdRequest)
+static esp_err_t RequestDataAccess(httpd_req_t *HttpdRequest)
```
Comment (Behn4m) [inline components/OAuth2.0/OauthHttpLocalServer.c:53]: hardcoded value in L60

Comment (Behn4m) [reply components/OAuth2.0/OauthHttpLocalServer.c:53]: I'm not sure if `302` is standard value or customizable status that should be configurable! (L63)

File: components/OAuth2.0/OauthHttpLocalServer.c [hunk 19]
```diff
@@ -1,19 +1,18 @@
...
  * @brief This function searches for specific patterns ('code' and 'state') within a character array and returns a boolean value indicating if either pattern was Found.
- * @param[in] Response The character array to search within, and Response is response from first stage from spotify athurisiation
+ * @param[in] Response The character array to search within, and Response is response from first stage from Service athurisiation
  * @param[in] SizeRes The size of the character array.
  * @return Returns true if either the 'code' or 'state' pattern was Found, and false otherwise.
  */
-bool Spotify_FindCode(char *Response, uint16_t SizeRes)
+bool FindCode(char *Response, uint16_t SizeRes)
 {
     char *codeString = {"code"};
```
Comment (Behn4m) [inline components/OAuth2.0/OauthHttpLocalServer.c:19]: is the code always the same for different Oauth providers? if not, it should be reconfigurable value

**Review by Behn4m — CHANGES_REQUESTED** [review]
Great job Mohadese, you are almost there. I added some comments for improvement and configurability, pleache checkout them

**Review by Behn4m — CHANGES_REQUESTED** [review]
Great Job Mohadese. You are almost there. There are some modifications needed to improve modularity and configurablity. please checkout my comments.

File: main/main.c (hunk 10 above)
Comment (m4github) [reply main/main.c]: done.

**Comment by m4github** [comment]
> Your program doesn't work, please fix it! Screenshot from 2024-07-07 11-24-44

fixed

File: components/OAuth2.0/OauthHttpLocalServer.c (hunk 19 above)
Comment (m4github) [reply components/OAuth2.0/OauthHttpLocalServer.c:19]: It just searches for the code and is always the same way

File: components/OAuth2.0/OauthHttpLocalServer.c (hunk 18 above)
Comment (m4github) [reply components/OAuth2.0/OauthHttpLocalServer.c:53]: _[302 Found](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302)

    This response code means that the URI of requested resource has been changed temporarily. Further changes in the URI might be made in the future. Therefore, this same URI should be used by the client in future requests._
it looks fine I think!

File: components/OAuth2.0/OauthHttpLocalServer.c (hunk 17 above)
Comment (m4github) [reply components/OAuth2.0/OauthHttpLocalServer.c]: done.

File: components/OAuth2.0/OauthHttpLocalServer.c (hunk 16 above)
Comment (m4github) [reply components/OAuth2.0/OauthHttpLocalServer.c]: done.

File: components/OAuth2.0/OauthHttpLocalServer.c (hunk 15 above)
Comment (m4github) [reply components/OAuth2.0/OauthHttpLocalServer.c]: done.

File: components/OAuth2.0/OauthHttpLocalServer.c (hunk 14 above)
Comment (m4github) [reply components/OAuth2.0/OauthHttpLocalServer.c]: done.

File: main/main.c [hunk 20]
```diff
@@ -2,72 +2,112 @@
...
+//TODO merge all in a static function
+    OAuthInterfaceHandler.IsServiceAuthorizedSemaphore = &IsSpotifyAuthorizedSemaphore;
+    OAuthInterfaceHandler.ConfigAddressInSpiffs = SpotifyConfigAddressInSpiffs;
+    Spotify_ClientInfo = Spotify_ClientConfigInit();
+    OAuthInterfaceHandler.ClientConfig.url = Spotify_ClientInfo.url;
+    OAuthInterfaceHandler.ClientConfig.host = Spotify_ClientInfo.host;
+    OAuthInterfaceHandler.ClientConfig.path = Spotify_ClientInfo.path;
+    OAuthInterfaceHandler.ClientConfig.requestURI = Spotify_ClientInfo.requestURI;
+    OAuthInterfaceHandler.ClientConfig.responseURI = Spotify_ClientInfo.responseURI;
+    OAuthInterfaceHandler.ClientConfig.hostname = Spotify_ClientInfo.hostname;
```
Comment (MohammadAbbasi76) [inline main/main.c]: what is different between "Spotify_ClientInfo.host" and "Spotify_ClientInfo.hostname" please choose other name,actually I'm confuse

File: components/OAuth2.0/authorization.c [hunk 21]
```diff
@@ -0,0 +1,423 @@
...
+
+/**
+ * @brief This function returns the access token.
+ * @param OAuthInterfaceHandler as the handler
+ * @return the access token
+ */
+char* Oauth_GetAccessToken(void)
+{
+    return AuthPrivateHandler.token.AccessToken;
+}
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: why you don't use queue?

File: main/main.c [hunk 22]
```diff
@@ -2,72 +2,112 @@
...
+    OAuthInterfaceHandler.ClientConfig.requestURL = Spotify_ClientInfo.requestURL; 
+    OAuthInterfaceHandler.ClientConfig.clientID = Spotify_ClientInfo.clientID;
+    OAuthInterfaceHandler.ClientConfig.base64Credintials = Spotify_ClientInfo.base64Credintials;
+    OAuthInterfaceHandler.ClientConfig.redirectURL = Spotify_ClientInfo.redirectURL;
+
+    Oauth_TaskInit(&OAuthInterfaceHandler);
```
Comment (MohammadAbbasi76) [inline main/main.c]: please check functions return

File: components/OAuth2.0/authorization.c [hunk 23]
```diff
@@ -0,0 +1,423 @@
...
+ * @param[in] Json The input JSON string.
+ * @param[in] token address to save tokin extracted information.
+ * @return false if fail, true if finish successful.
+ */
+static bool ExtractTokenJson(char *Json, Token_t *token) 
+{
+    const char *TAG = "JsonExTraction";
+    cJSON *J_Token = cJSON_Parse(Json);
+
+    if (J_Token == NULL)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: Please do all extraction in json.c file! I told you before

File: components/OAuth2.0/authorization.c [hunk 24]
```diff
@@ -0,0 +1,423 @@
...
+
+    return Request_Access_URI;
+}
+
+/**
+ * @brief Initialize and return an httpd_uri_t structure for the response URI.
+ * @param none
+ * @return An initialized httpd_uri_t structure with the specified URI, method, and handler.
+ */
+static httpd_uri_t ResponseURIInit(void)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: please replace it in to the OauthHttpLocalServer.c

File: components/OAuth2.0/authorization.c [hunk 25]
```diff
@@ -0,0 +1,423 @@
...
+    ESP_LOGI(TAG, "the CODE found in response");
+    return ESP_OK;
+}
+
+/**
+ * @brief Initialize and return an httpd_uri_t structure for the request URI.
+ * @param none
+ * @return An initialized httpd_uri_t structure with the specified URI, method, and handler.
+*/
+static httpd_uri_t RequestURIInit(void)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: please replace it in to the OauthHttpLocalServer.c

File: components/OAuth2.0/authorization.c [hunk 26]
```diff
@@ -0,0 +1,423 @@
...
+    free(localURL);
+    return ESP_OK;
+}
+
+/**
+ * @brief This function handles the callback HTTPS request from Service and processes the response data.
+ * @param[in] req The HTTP request object.
+ * @return Returns ESP_OK if the request is processed successfully.
+ */
+static esp_err_t HttpsCallbackHandler(httpd_req_t *HttpdRequest)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: please replace it in to the OauthHttpLocalServer.c

File: components/OAuth2.0/authorization.c [hunk 27]
```diff
@@ -0,0 +1,423 @@
...
+    return false; // Access token substring not Found
+}
+
+/**
+ * @brief This function handles the first HTTPS request to Service and 
+ *        redirects the user to the authorization page.
+ * @param[in] req The HTTP request object.
+ * @return Returns ESP_OK if the request is processed successfully.
+ */
+static esp_err_t RequestDataAccess(httpd_req_t *HttpdRequest)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: please replace it in to the OauthHttpLocalServer.c

File: components/OAuth2.0/authorization.c [hunk 28]
```diff
@@ -0,0 +1,423 @@
...
+
+/**
+ * @brief This function searches for specific patterns ('code' and 'state') within a character 
+ *        array and returns a boolean value indicating if either pattern was Found.
+ * @param[in] Response The character array to search within, 
+ *            and Response is response from first stage from Service athurisiation
+ * @param[in] SizeRes The size of the character array.
+ * @return Returns true if either the 'code' or 'state' pattern was Found, and false otherwise.
+ */
+static bool FindCode(char *Response, uint16_t SizeRes)
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/authorization.c]: please replace it in to the OauthAPICall.c

**Comment by MohammadAbbasi76** [comment]
please use same way for naming for authorization.c
Screenshot from 2024-07-17 11-35-56

File: components/OAuth2.0/OauthAPICall.c [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+    esp_http_client_set_header(httpClient, "Cookie", "__Host-device_id=AQAwmp7jxagopcWw89BjSDAA530mHwIieOZdJ9Im8nI0-70oEsSInx3jkeSO09YQ7sPgPaIUyMEvZ-tct7I6OlshJrzVYOqcgo0; sp_tr=false");
+
+    // Set the request body (POST data)
+    char Grand[MEDIUM_BUF] = {0};    
+    sprintf(Grand, "grant_type=authorization_code&redirect_uri=%s&%s", ClientConfig->redirectURL, Code);
+    esp_http_client_set_post_field(httpClient, Grand, strlen(Grand));
+
+    // Enable detailed logging for debugging
+    // esp_http_client_set_debug(client, true);
+
```
Comment (MohammadAbbasi76) [inline components/OAuth2.0/OauthAPICall.c]: please remove all extera comments

File: components/OAuth2.0/authorization.c (hunk 11 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: fixed.

File: components/OAuth2.0/authorization.c (hunk 12 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: fixed.

File: components/OAuth2.0/authorization.c (hunk 13 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: fixed.

File: components/SpotifyInterface/SpotifyInterface.c [hunk 30]
```diff
@@ -1,289 +1,246 @@
...
+static esp_err_t HttpEventHandler(esp_http_client_event_t *evt) 
+{
+    static int totalLen = 0;
+    // Create the queue and semaphore
 
-        ESP_LOGI(TAG, "Spotify app initiated successfully");
-    }
-    else
+    switch (evt->event_id) 
     {
```
Comment (MohammadAbbasi76) [inline components/SpotifyInterface/SpotifyInterface.c]: what is this, why replace it?

File: components/OAuth2.0/authorization.c (hunk 21 above)
Comment (Behn4m) [reply components/OAuth2.0/authorization.c]: this function just reads the memory and returns the value. No wait or inter-thread communication is needed

File: components/OAuth2.0/authorization.c (hunk 23 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: It's a static function called just for ones so I prefer to keep it here in *.c file...It looks more optimal this way.

File: components/OAuth2.0/authorization.c (hunk 26 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: done.

File: components/OAuth2.0/authorization.c (hunk 27 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: done.

File: components/OAuth2.0/authorization.c (hunk 28 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: done.

File: components/OAuth2.0/authorization.c (hunk 25 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: done.

File: components/OAuth2.0/authorization.c (hunk 24 above)
Comment (m4github) [reply components/OAuth2.0/authorization.c]: done.

**Comment by m4github** [comment]
> please use same way for naming for authorization.c Screenshot from 2024-07-17 11-35-56

done.

File: components/OAuth2.0/OauthAPICall.c (hunk 29 above)
Comment (m4github) [reply components/OAuth2.0/OauthAPICall.c]: done.

File: components/SpotifyInterface/SpotifyInterface.c (hunk 30 above)
Comment (m4github) [reply components/SpotifyInterface/SpotifyInterface.c]: you mean the log? cause it won't init here anymore.

File: main/main.c (hunk 22 above)
Comment (m4github) [reply main/main.c]: done.

File: main/main.c (hunk 20 above)
Comment (m4github) [reply main/main.c]: the hostname used for MDNS stuff...fixed.

File: components/OAuth2.0/include/OauthAPICall.h [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+void APICallInit(APIBuffer_t *APIBuffer, char *base64Cred);
+
+/**
+ * @brief  This function sends a request to the Service login API 
+ *         to exchange an authorization code for an access token.
+ * @param  RefreshToken The refresh token.
+ * @param  ClientConfig The client configuration.
+ * @return This function does not return a value.
+ */
+void SendRequest_ExchangeTokenWithRefreshToken(char *RefreshToken, HttpClientInfo_t *ClientConfig);
```
Comment (Behn4m) [inline components/OAuth2.0/include/OauthAPICall.h]: drop `SendRequest` to match naming of rest of functions

Comment (Behn4m) [reply components/OAuth2.0/include/OauthAPICall.h]: or `addOauth_` to the beginning of all functions nams

File: components/SpotifyInterface/SpotifyInterface.c [hunk 32]
```diff
@@ -1,289 +1,246 @@
 #include <stdio.h>
-#include "SpotifyTypedef.h"
+#include <stdlib.h>
+#include <string.h>
+//#include "cJSON.h"
```
Comment (Behn4m) [inline components/SpotifyInterface/SpotifyInterface.c]: remove this line if not needed

File: components/SpotifyInterface/include/SpotifyInterface.h [hunk 33]
```diff
@@ -67,43 +67,53 @@ typedef struct PlaybackInfo_t
...
+    char *redirectURL;
+} SpotifyHttpInfo_t;
 
 /**
  * @brief This function deinitiates the Spotify authorization process.
  * @param SpotifyInterfaceHandler as the handler
  * @return true if task run to the end
  */
-void Spotify_TaskDeinit(SpotifyInterfaceHandler_t *SpotifyInterfaceHandler);
+//void Spotify_TaskDeinit(SpotifyInterfaceHandler_t *SpotifyInterfaceHandler);
```
Comment (Behn4m) [inline components/SpotifyInterface/include/SpotifyInterface.h]: remove if not needed

File: components/SpotifyInterface/SpotifyInterface.c (hunk 32 above)
Comment (m4github) [reply components/SpotifyInterface/SpotifyInterface.c]: done.

File: components/SpotifyInterface/include/SpotifyInterface.h (hunk 33 above)
Comment (m4github) [reply components/SpotifyInterface/include/SpotifyInterface.h]: done.

File: components/OAuth2.0/include/OauthAPICall.h (hunk 31 above)
Comment (m4github) [reply components/OAuth2.0/include/OauthAPICall.h]: done.

**Review by MohammadAbbasi76 — APPROVED** [review]

**Review by Behn4m — APPROVED** [review]

**Pull request merged**